File: docs/android-widget/java/MetaboleWidget.java
package app.metabole.client;

// Widget da home screen di Metabole (Java: il progetto Capacitor è Java-only).
// Legge il "token widget" salvato dall'app in SharedPreferences ("CapacitorStorage",
// chiave "metabole_widget_token"), chiama GET /widget e mostra la mascotte Gaia con
// stato/saluto/frase/pasto/acqua/passi/streak in TRE formati (quadrato, rettangolare, largo).
// Tap → apre l'app.

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.SizeF;
import android.widget.RemoteViews;

import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

public class MetaboleWidget extends AppWidgetProvider {

    private static final String API = "https://metabole-backend.onrender.com/api/v1";

    @Override
    public void onUpdate(Context context, AppWidgetManager mgr, int[] ids) {
        for (int id : ids) apply(context, mgr, id, null, "Aggiorno…");
        new Thread(() -> {
            JSONObject data = null;
            try { data = load(context); } catch (Exception ignored) {}
            final JSONObject d = data;
            for (int id : ids) apply(context, mgr, id, d, d == null ? "Apri l'app Metabole" : null);
        }).start();
    }

    @Override
    public void onAppWidgetOptionsChanged(Context context, AppWidgetManager mgr, int id, Bundle newOptions) {
        // Su ridimensionamento ricarichiamo (il mapping per formato viene ricalcolato).
        new Thread(() -> {
            JSONObject data = null;
            try { data = load(context); } catch (Exception ignored) {}
            apply(context, mgr, id, data, data == null ? "Apri l'app Metabole" : null);
        }).start();
    }

    private JSONObject load(Context context) throws Exception {
        String token = context
                .getSharedPreferences("CapacitorStorage", Context.MODE_PRIVATE)
                .getString("metabole_widget_token", null);
        if (token == null || token.isEmpty()) return null;
        HttpURLConnection conn = (HttpURLConnection) new URL(API + "/widget").openConnection();
        try {
            conn.setRequestMethod("GET");
            conn.setConnectTimeout(8000);
            conn.setReadTimeout(8000);
            conn.setRequestProperty("Authorization", "Bearer " + token);
            int code = conn.getResponseCode();
            if (code < 200 || code > 299) return null;
            BufferedReader r = new BufferedReader(new InputStreamReader(conn.getInputStream()));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = r.readLine()) != null) sb.append(line);
            r.close();
            return new JSONObject(sb.toString());
        } finally {
            conn.disconnect();
        }
    }

    // Applica la vista giusta in base alle dimensioni disponibili.
    private void apply(Context context, AppWidgetManager mgr, int id, JSONObject d, String placeholder) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            Map<SizeF, RemoteViews> views = new HashMap<>();
            views.put(new SizeF(120f, 120f), build(context, R.layout.widget_square, d, placeholder));
            views.put(new SizeF(220f, 120f), build(context, R.layout.widget_rect, d, placeholder));
            views.put(new SizeF(250f, 220f), build(context, R.layout.widget_large, d, placeholder));
            mgr.updateAppWidget(id, new RemoteViews(views));
        } else {
            Bundle opt = mgr.getAppWidgetOptions(id);
            int minW = opt != null ? opt.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_WIDTH, 0) : 0;
            int minH = opt != null ? opt.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_HEIGHT, 0) : 0;
            int layout;
            if (minH >= 200 && minW >= 240) layout = R.layout.widget_large;
            else if (minW >= 200) layout = R.layout.widget_rect;
            else layout = R.layout.widget_square;
            mgr.updateAppWidget(id, build(context, layout, d, placeholder));
        }
    }

    private RemoteViews build(Context context, int layout, JSONObject d, String placeholder) {
        RemoteViews v = new RemoteViews(context.getPackageName(), layout);

        // Tap sull'intero widget → apre l'app.
        Intent launch = context.getPackageManager().getLaunchIntentForPackage(context.getPackageName());
        if (launch != null) {
            PendingIntent pi = PendingIntent.getActivity(context, 0, launch,
                    PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
            v.setOnClickPendingIntent(R.id.widget_root, pi);
        }

        String state = d != null ? d.optString("state", "inrotta") : "inrotta";
        v.setInt(R.id.widget_root, "setBackgroundResource", bgFor(state));
        setImage(v, R.id.widget_mascot, "buonanotte".equals(state) ? R.drawable.mascot_sleepy : R.drawable.mascot_happy);

        if (d != null) {
            setText(v, R.id.widget_greeting, d.optString("greeting", "Metabole"));
            setText(v, R.id.widget_phrase, "“" + d.optString("phrase", "") + "”");
            setText(v, R.id.widget_streak, "🔥 " + d.optInt("streak", 0) + " giorni");

            JSONObject water = d.optJSONObject("water");
            JSONObject steps = d.optJSONObject("steps");
            // Unità acqua scelta dal cliente: icona + valore come in dashboard.
            String waterUnit = d.optString("waterUnit", "glass");
            int perGlasses = waterUnitGlasses(waterUnit);
            setText(v, R.id.widget_water, water != null
                    ? waterVal(water.optInt("glasses"), perGlasses) + "/" + waterVal(water.optInt("goal"), perGlasses) : "0/8");
            setText(v, R.id.widget_water_icon, waterEmoji(waterUnit));
            setText(v, R.id.widget_steps, steps != null ? thousands(steps.optInt("steps")) : "0");

            JSONObject meal = d.optJSONObject("nextMeal");
            if (meal != null) {
                setText(v, R.id.widget_meal_label, "Prossimo pasto · " + slotLabel(meal.optString("slot")));
                setText(v, R.id.widget_meal_name, meal.optString("name"));
                setText(v, R.id.widget_meal_kcal, meal.optInt("kcal") + " kcal");
            } else {
                setText(v, R.id.widget_meal_label, "Prossimo pasto");
                setText(v, R.id.widget_meal_name, "Nessun pasto in programma");
                setText(v, R.id.widget_meal_kcal, "");
            }
        } else {
            setText(v, R.id.widget_greeting, "Metabole");
            setText(v, R.id.widget_phrase, placeholder != null ? placeholder : "");
            setText(v, R.id.widget_streak, "🔥 0");
            setText(v, R.id.widget_water, "0/8");
            setText(v, R.id.widget_steps, "0");
            setText(v, R.id.widget_meal_label, "Prossimo pasto");
            setText(v, R.id.widget_meal_name, "Apri l'app Metabole");
            setText(v, R.id.widget_meal_kcal, "");
        }
        return v;
    }

    // setTextViewText ma tollerante ai layout che non hanno quel campo (id assente → no-op sicuro).
    private void setText(RemoteViews v, int id, String text) {
        try { v.setTextViewText(id, text); } catch (Exception ignored) {}
    }

    private void setImage(RemoteViews v, int id, int res) {
        try { v.setImageViewResource(id, res); } catch (Exception ignored) {}
    }

    private int bgFor(String state) {
        switch (state) {
            case "buongiorno":
            case "acqua":
                return R.drawable.widget_bg_blue;
            case "passi":
                return R.drawable.widget_bg_orange;
            case "buonanotte":
                return R.drawable.widget_bg_dark;
            default:
                return R.drawable.widget_bg_teal;
        }
    }

    private String slotLabel(String slot) {
        switch (slot) {
            case "breakfast": return "Colazione";
            case "morning_snack": return "Spuntino";
            case "lunch": return "Pranzo";
            case "afternoon_snack": return "Merenda";
            case "dinner": return "Cena";
            default: return "Pasto";
        }
    }

    private String thousands(int n) {
        String s = Integer.toString(n);
        StringBuilder out = new StringBuilder();
        int c = 0;
        for (int i = s.length() - 1; i >= 0; i--) {
            out.insert(0, s.charAt(i));
            if (++c % 3 == 0 && i > 0) out.insert(0, '.');
        }
        return out.toString();
    }

    // ---- Unità acqua (stessa logica di app/src/lib/water.ts): 1 bicchiere = 250 ml.
    // glass=1, bottle05=2, bottle1=4, bottle15=6 bicchieri per unità.
    private int waterUnitGlasses(String unit) {
        switch (unit) {
            case "bottle05": return 2;
            case "bottle1": return 4;
            case "bottle15": return 6;
            default: return 1; // glass
        }
    }

    private String waterEmoji(String unit) {
        return "glass".equals(unit) ? "💧" : "🍶"; // goccia per i bicchieri, bottiglia per le bottiglie
    }

    // Converte i bicchieri nell'unità scelta: intero se tondo, altrimenti 1 decimale (virgola).
    private String waterVal(int glasses, int per) {
        if (per <= 1) return String.valueOf(glasses);
        double vv = (double) glasses / per;
        if (vv == Math.floor(vv)) return String.valueOf((int) vv);
        return String.format(java.util.Locale.ITALY, "%.1f", vv);
    }
}

File: docs/android-widget/MetaboleWidget.kt
package app.metabole.client

// ⚠️ Il package sopra DEVE combaciare con quello della MainActivity del progetto
// (guarda android/app/src/main/java/.../MainActivity.java). Se è diverso, correggilo.

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.widget.RemoteViews
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Widget da home screen di Metabole.
 * Legge il "token widget" salvato dall'app in SharedPreferences ("CapacitorStorage",
 * chiave "metabole_widget_token"), chiama GET /widget e mostra stato mascotte,
 * saluto, frase, prossimo pasto e acqua/passi. Tap → apre l'app.
 */
class MetaboleWidget : AppWidgetProvider() {

    override fun onUpdate(context: Context, mgr: AppWidgetManager, ids: IntArray) {
        // Stato immediato (evita widget vuoto mentre carica)
        for (id in ids) render(context, mgr, id, null, "Aggiorno…")
        Thread {
            val data = try { load(context) } catch (e: Exception) { null }
            for (id in ids) {
                if (data != null) render(context, mgr, id, data, null)
                else render(context, mgr, id, null, "Apri l'app Metabole")
            }
        }.start()
    }

    private fun load(context: Context): JSONObject? {
        val token = context
            .getSharedPreferences("CapacitorStorage", Context.MODE_PRIVATE)
            .getString("metabole_widget_token", null) ?: return null
        val conn = URL("$API/widget").openConnection() as HttpURLConnection
        return try {
            conn.requestMethod = "GET"
            conn.connectTimeout = 8000
            conn.readTimeout = 8000
            conn.setRequestProperty("Authorization", "Bearer $token")
            if (conn.responseCode in 200..299)
                JSONObject(conn.inputStream.bufferedReader().readText())
            else null
        } finally {
            conn.disconnect()
        }
    }

    private fun render(
        context: Context,
        mgr: AppWidgetManager,
        id: Int,
        d: JSONObject?,
        placeholder: String?,
    ) {
        val v = RemoteViews(context.packageName, R.layout.metabole_widget)

        // Tap sul widget → apre l'app
        val launch = context.packageManager.getLaunchIntentForPackage(context.packageName)
        if (launch != null) {
            val pi = PendingIntent.getActivity(
                context, 0, launch,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
            )
            v.setOnClickPendingIntent(R.id.widget_root, pi)
        }

        if (d != null) {
            val emoji = when (d.optString("state")) {
                "buongiorno" -> "🌅"
                "acqua" -> "💧"
                "passi" -> "🚶"
                "buonanotte" -> "🌙"
                else -> "🌱"
            }
            v.setTextViewText(R.id.widget_emoji, emoji)
            v.setTextViewText(R.id.widget_greeting, d.optString("greeting", "Metabole"))
            v.setTextViewText(R.id.widget_phrase, d.optString("phrase", ""))

            val meal = d.optJSONObject("nextMeal")
            v.setTextViewText(
                R.id.widget_meal,
                if (meal != null) "🍽 ${meal.optString("name")} · ${meal.optInt("kcal")} kcal"
                else "Nessun pasto in programma",
            )

            val water = d.optJSONObject("water")
            val steps = d.optJSONObject("steps")
            val wt = if (water != null) "💧 ${water.optInt("glasses")}/${water.optInt("goal")}" else ""
            val st = if (steps != null) "🚶 ${steps.optInt("steps")}" else ""
            v.setTextViewText(R.id.widget_metrics, "$wt   $st")
        } else {
            v.setTextViewText(R.id.widget_emoji, "🌱")
            v.setTextViewText(R.id.widget_greeting, "Metabole")
            v.setTextViewText(R.id.widget_phrase, placeholder ?: "")
            v.setTextViewText(R.id.widget_meal, "")
            v.setTextViewText(R.id.widget_metrics, "")
        }

        mgr.updateAppWidget(id, v)
    }

    companion object {
        // Se un giorno cambi backend, aggiorna qui.
        private const val API = "https://metabole-backend.onrender.com/api/v1"
    }
}
